File: go/indicators/johnehlers/coronaswingposition/coronaswingposition.go
// Package coronaswingposition implements Ehlers' Corona Swing Position heatmap
// indicator.
//
// The Corona swing position is estimated by correlating the prices with a
// perfect sine wave having the dominant cycle period. This correlation produces
// a smooth waveform that lets us better estimate the swing position and
// impending turning points.
//
// Reference: John Ehlers, "Measuring Cycle Periods", Stocks & Commodities,
// November 2008.
package coronaswingposition

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
	"zpano/indicators/johnehlers/corona"
)

const (
	maxLeadListCount     = 50
	maxPositionListCount = 20

	// 60° phase-lead coefficients: lead60 = 0.5·BP + sin(60°)·Q ≈ 0.5·BP + 0.866·Q.
	lead60CoefBP = 0.5
	lead60CoefQ  = 0.866

	// Bandpass filter delta factor in γ = 1/cos(ω · 2 · 0.1).
	bpDelta = 0.1

	widthHighThreshold = 0.85
	widthHighSaturate  = 0.8
	widthNarrow        = 0.01
	widthScale         = 0.15

	rasterBlendExponent = 0.95
	rasterBlendHalf     = 0.5
)

// CoronaSwingPosition is Ehlers' Corona Swing Position indicator.
type CoronaSwingPosition struct {
	mu                  sync.RWMutex
	mnemonic            string
	description         string
	mnemonicSP          string
	descriptionSP       string
	c                   *corona.Corona
	rasterLength        int
	rasterStep          float64
	maxRasterValue      float64
	minParameterValue   float64
	maxParameterValue   float64
	parameterResolution float64
	raster              []float64
	leadList            []float64
	positionList        []float64
	samplePrevious      float64
	samplePrevious2     float64
	bandPassPrevious    float64
	bandPassPrevious2   float64
	swingPosition       float64
	isStarted           bool
	barFunc             entities.BarFunc
	quoteFunc           entities.QuoteFunc
	tradeFunc           entities.TradeFunc
}

// NewCoronaSwingPositionDefault returns an instance created with default parameters.
func NewCoronaSwingPositionDefault() (*CoronaSwingPosition, error) {
	return NewCoronaSwingPositionParams(&Params{})
}

// NewCoronaSwingPositionParams returns an instance created with the supplied parameters.
//
//nolint:funlen,cyclop
func NewCoronaSwingPositionParams(p *Params) (*CoronaSwingPosition, error) {
	const (
		invalid      = "invalid corona swing position parameters"
		fmtRaster    = "%s: RasterLength should be >= 2"
		fmtMaxRaster = "%s: MaxRasterValue should be > 0"
		fmtMaxParam  = "%s: MaxParameterValue should be > MinParameterValue"
		fmtHP        = "%s: HighPassFilterCutoff should be >= 2"
		fmtMinP      = "%s: MinimalPeriod should be >= 2"
		fmtMaxP      = "%s: MaximalPeriod should be > MinimalPeriod"
		fmtw         = "%s: %w"
		fmtnValue    = "cswing(%d, %g, %g, %g, %d%s)"
		fmtnSP       = "cswing-sp(%d%s)"
		descrValue   = "Corona swing position "
		descrSP      = "Corona swing position scalar "
		defRaster    = 50
		defMaxRast   = 20.0
		defMinParam  = -5.0
		defMaxParam  = 5.0
		defHPCutoff  = 30
		defMinPer    = 6
		defMaxPer    = 30
	)

	cfg := *p

	if cfg.RasterLength == 0 {
		cfg.RasterLength = defRaster
	}

	if cfg.MaxRasterValue == 0 {
		cfg.MaxRasterValue = defMaxRast
	}

	// MinParameterValue and MaxParameterValue default to -5 and 5; since 0 is a
	// valid user value for either, we only substitute when both are zero (the
	// unconfigured case).
	if cfg.MinParameterValue == 0 && cfg.MaxParameterValue == 0 {
		cfg.MinParameterValue = defMinParam
		cfg.MaxParameterValue = defMaxParam
	}

	if cfg.HighPassFilterCutoff == 0 {
		cfg.HighPassFilterCutoff = defHPCutoff
	}

	if cfg.MinimalPeriod == 0 {
		cfg.MinimalPeriod = defMinPer
	}

	if cfg.MaximalPeriod == 0 {
		cfg.MaximalPeriod = defMaxPer
	}

	if cfg.RasterLength < 2 {
		return nil, fmt.Errorf(fmtRaster, invalid)
	}

	if cfg.MaxRasterValue <= 0 {
		return nil, fmt.Errorf(fmtMaxRaster, invalid)
	}

	if cfg.MaxParameterValue <= cfg.MinParameterValue {
		return nil, fmt.Errorf(fmtMaxParam, invalid)
	}

	if cfg.HighPassFilterCutoff < 2 {
		return nil, fmt.Errorf(fmtHP, invalid)
	}

	if cfg.MinimalPeriod < 2 {
		return nil, fmt.Errorf(fmtMinP, invalid)
	}

	if cfg.MaximalPeriod <= cfg.MinimalPeriod {
		return nil, fmt.Errorf(fmtMaxP, invalid)
	}

	bc := cfg.BarComponent
	if bc == 0 {
		bc = entities.BarMedianPrice
	}

	qc := cfg.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := cfg.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	c, err := corona.NewCorona(&corona.Params{
		HighPassFilterCutoff: cfg.HighPassFilterCutoff,
		MinimalPeriod:        cfg.MinimalPeriod,
		MaximalPeriod:        cfg.MaximalPeriod,
	})
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	barFunc, err := entities.BarComponentFunc(bc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	quoteFunc, err := entities.QuoteComponentFunc(qc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	tradeFunc, err := entities.TradeComponentFunc(tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	componentMnemonic := core.ComponentTripleMnemonic(bc, qc, tc)

	mnemonicValue := fmt.Sprintf(fmtnValue,
		cfg.RasterLength, cfg.MaxRasterValue, cfg.MinParameterValue, cfg.MaxParameterValue,
		cfg.HighPassFilterCutoff, componentMnemonic)
	mnemonicSP := fmt.Sprintf(fmtnSP, cfg.HighPassFilterCutoff, componentMnemonic)

	parameterResolution := float64(cfg.RasterLength-1) / (cfg.MaxParameterValue - cfg.MinParameterValue)

	return &CoronaSwingPosition{
		mnemonic:            mnemonicValue,
		description:         descrValue + mnemonicValue,
		mnemonicSP:          mnemonicSP,
		descriptionSP:       descrSP + mnemonicSP,
		c:                   c,
		rasterLength:        cfg.RasterLength,
		rasterStep:          cfg.MaxRasterValue / float64(cfg.RasterLength),
		maxRasterValue:      cfg.MaxRasterValue,
		minParameterValue:   cfg.MinParameterValue,
		maxParameterValue:   cfg.MaxParameterValue,
		parameterResolution: parameterResolution,
		raster:              make([]float64, cfg.RasterLength),
		leadList:            make([]float64, 0, maxLeadListCount),
		positionList:        make([]float64, 0, maxPositionListCount),
		swingPosition:       math.NaN(),
		barFunc:             barFunc,
		quoteFunc:           quoteFunc,
		tradeFunc:           tradeFunc,
	}, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *CoronaSwingPosition) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.c.IsPrimed()
}

// Metadata describes the output data of the indicator.
func (s *CoronaSwingPosition) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.CoronaSwingPosition,
		s.mnemonic,
		s.description,
		[]core.OutputText{
			{Mnemonic: s.mnemonic, Description: s.description},
			{Mnemonic: s.mnemonicSP, Description: s.descriptionSP},
		},
	)
}

// Update feeds the next sample and returns the heatmap column plus the current
// SwingPosition. On unprimed bars the heatmap is empty and the scalar is NaN.
//
//nolint:funlen,cyclop,gocognit
func (s *CoronaSwingPosition) Update(sample float64, t time.Time) (*outputs.Heatmap, float64) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if math.IsNaN(sample) {
		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution), math.NaN()
	}

	primed := s.c.Update(sample)

	if !s.isStarted {
		s.samplePrevious = sample
		s.isStarted = true

		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution), math.NaN()
	}

	// Bandpass InPhase filter at the dominant cycle median period.
	omega := 2. * math.Pi / s.c.DominantCycleMedian()
	beta2 := math.Cos(omega)
	gamma2 := 1. / math.Cos(omega*2*bpDelta)
	alpha2 := gamma2 - math.Sqrt(gamma2*gamma2-1.)
	bandPass := 0.5*(1-alpha2)*(sample-s.samplePrevious2) +
		beta2*(1+alpha2)*s.bandPassPrevious -
		alpha2*s.bandPassPrevious2

	// Quadrature = derivative / omega.
	quadrature2 := (bandPass - s.bandPassPrevious) / omega

	s.bandPassPrevious2 = s.bandPassPrevious
	s.bandPassPrevious = bandPass
	s.samplePrevious2 = s.samplePrevious
	s.samplePrevious = sample

	// 60° lead: lead60 = 0.5·BP_previous2 + 0.866·Q
	lead60 := lead60CoefBP*s.bandPassPrevious2 + lead60CoefQ*quadrature2

	lowest, highest := appendRolling(&s.leadList, maxLeadListCount, lead60)

	// Normalised lead position ∈ [0, 1].
	position := highest - lowest
	if position > 0 {
		position = (lead60 - lowest) / position
	}

	lowest, highest = appendRolling(&s.positionList, maxPositionListCount, position)
	highest -= lowest

	width := 0.15 * highest
	if highest > widthHighThreshold {
		width = widthNarrow
	}

	s.swingPosition = (s.maxParameterValue-s.minParameterValue)*position + s.minParameterValue

	positionScaledToRasterLength := int(math.Round(position * float64(s.rasterLength)))
	positionScaledToMaxRasterValue := position * s.maxRasterValue

	for i := 0; i < s.rasterLength; i++ {
		value := s.raster[i]

		if i == positionScaledToRasterLength {
			value *= rasterBlendHalf
		} else {
			argument := positionScaledToMaxRasterValue - s.rasterStep*float64(i)
			if i > positionScaledToRasterLength {
				argument = -argument
			}

			if width > 0 {
				value = rasterBlendHalf *
					(math.Pow(argument/width, rasterBlendExponent) + rasterBlendHalf*value)
			}
		}

		if value < 0 {
			value = 0
		} else if value > s.maxRasterValue {
			value = s.maxRasterValue
		}

		if highest > widthHighSaturate {
			value = s.maxRasterValue
		}

		if math.IsNaN(value) {
			value = 0
		}

		s.raster[i] = value
	}

	if !primed {
		return outputs.NewEmptyHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution), math.NaN()
	}

	values := make([]float64, s.rasterLength)
	valueMin := math.Inf(1)
	valueMax := math.Inf(-1)

	for i := 0; i < s.rasterLength; i++ {
		v := s.raster[i]
		values[i] = v

		if v < valueMin {
			valueMin = v
		}

		if v > valueMax {
			valueMax = v
		}
	}

	heatmap := outputs.NewHeatmap(t, s.minParameterValue, s.maxParameterValue, s.parameterResolution,
		valueMin, valueMax, values)

	return heatmap, s.swingPosition
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *CoronaSwingPosition) UpdateScalar(sample *entities.Scalar) core.Output {
	return s.updateEntity(sample.Time, sample.Value)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *CoronaSwingPosition) UpdateBar(sample *entities.Bar) core.Output {
	return s.updateEntity(sample.Time, s.barFunc(sample))
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *CoronaSwingPosition) UpdateQuote(sample *entities.Quote) core.Output {
	return s.updateEntity(sample.Time, s.quoteFunc(sample))
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *CoronaSwingPosition) UpdateTrade(sample *entities.Trade) core.Output {
	return s.updateEntity(sample.Time, s.tradeFunc(sample))
}

func (s *CoronaSwingPosition) updateEntity(t time.Time, sample float64) core.Output {
	const length = 2

	heatmap, sp := s.Update(sample, t)

	output := make([]any, length)
	output[0] = heatmap
	output[1] = entities.Scalar{Time: t, Value: sp}

	return output
}

// appendRolling appends v to the list, drops the oldest element once the list
// reaches maxCount, and returns the current (lowest, highest) values.
func appendRolling(list *[]float64, maxCount int, v float64) (float64, float64) {
	if len(*list) >= maxCount {
		*list = (*list)[1:]
	}

	*list = append(*list, v)

	lowest := v
	highest := v

	for _, x := range *list {
		if x < lowest {
			lowest = x
		}

		if x > highest {
			highest = x
		}
	}

	return lowest, highest
}

File: ts/indicators/john-ehlers/corona-swing-position/corona-swing-position.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { Heatmap } from '../../core/outputs/heatmap';
import { Corona } from '../corona/corona';
import { CoronaSwingPositionParams } from './params';

const DEFAULT_RASTER_LENGTH = 50;
const DEFAULT_MAX_RASTER = 20;
const DEFAULT_MIN_PARAM = -5;
const DEFAULT_MAX_PARAM = 5;
const DEFAULT_HP_CUTOFF = 30;
const DEFAULT_MIN_PERIOD = 6;
const DEFAULT_MAX_PERIOD = 30;

const MAX_LEAD_LIST_COUNT = 50;
const MAX_POSITION_LIST_COUNT = 20;

// 60° phase-lead coefficients.
const LEAD60_COEF_BP = 0.5;
const LEAD60_COEF_Q = 0.866;

const BP_DELTA = 0.1;

const WIDTH_HIGH_THRESHOLD = 0.85;
const WIDTH_HIGH_SATURATE = 0.8;
const WIDTH_NARROW = 0.01;

const RASTER_BLEND_EXPONENT = 0.95;
const RASTER_BLEND_HALF = 0.5;

/** __Corona Swing Position__ (Ehlers) heatmap indicator.
 *
 * Correlates prices with a perfect sine wave having the dominant cycle period, producing a
 * smooth waveform that lets us better estimate the swing position and impending turning points.
 *
 * Reference: John Ehlers, "Measuring Cycle Periods", Stocks & Commodities, November 2008.
 */
export class CoronaSwingPosition implements Indicator {
  private readonly c: Corona;
  private readonly rasterLength: number;
  private readonly rasterStep: number;
  private readonly maxRasterValue: number;
  private readonly minParameterValue: number;
  private readonly maxParameterValue: number;
  private readonly parameterResolution: number;
  private readonly raster: number[];
  private readonly leadList: number[] = [];
  private readonly positionList: number[] = [];

  private readonly mnemonicValue: string;
  private readonly descriptionValue: string;
  private readonly mnemonicSP: string;
  private readonly descriptionSP: string;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  private samplePrevious = 0;
  private samplePrevious2 = 0;
  private bandPassPrevious = 0;
  private bandPassPrevious2 = 0;
  private swingPosition = Number.NaN;
  private isStarted = false;

  public static default(): CoronaSwingPosition {
    return new CoronaSwingPosition({});
  }

  public static fromParams(params: CoronaSwingPositionParams): CoronaSwingPosition {
    return new CoronaSwingPosition(params);
  }

  private constructor(params: CoronaSwingPositionParams) {
    const invalid = 'invalid corona swing position parameters';

    const rasterLength = params.rasterLength !== undefined && params.rasterLength !== 0
      ? params.rasterLength : DEFAULT_RASTER_LENGTH;
    const maxRaster = params.maxRasterValue !== undefined && params.maxRasterValue !== 0
      ? params.maxRasterValue : DEFAULT_MAX_RASTER;

    // Only substitute Min/Max when both are 0 (unconfigured), since 0 is a valid user value.
    let minParam = params.minParameterValue ?? 0;
    let maxParam = params.maxParameterValue ?? 0;
    if (minParam === 0 && maxParam === 0) {
      minParam = DEFAULT_MIN_PARAM;
      maxParam = DEFAULT_MAX_PARAM;
    }

    const hpCutoff = params.highPassFilterCutoff !== undefined && params.highPassFilterCutoff !== 0
      ? params.highPassFilterCutoff : DEFAULT_HP_CUTOFF;
    const minPeriod = params.minimalPeriod !== undefined && params.minimalPeriod !== 0
      ? params.minimalPeriod : DEFAULT_MIN_PERIOD;
    const maxPeriod = params.maximalPeriod !== undefined && params.maximalPeriod !== 0
      ? params.maximalPeriod : DEFAULT_MAX_PERIOD;

    if (rasterLength < 2) throw new Error(`${invalid}: RasterLength should be >= 2`);
    if (maxRaster <= 0) throw new Error(`${invalid}: MaxRasterValue should be > 0`);
    if (maxParam <= minParam) throw new Error(`${invalid}: MaxParameterValue should be > MinParameterValue`);
    if (hpCutoff < 2) throw new Error(`${invalid}: HighPassFilterCutoff should be >= 2`);
    if (minPeriod < 2) throw new Error(`${invalid}: MinimalPeriod should be >= 2`);
    if (maxPeriod <= minPeriod) throw new Error(`${invalid}: MaximalPeriod should be > MinimalPeriod`);

    const bc = params.barComponent ?? BarComponent.Median;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    this.c = new Corona({
      highPassFilterCutoff: hpCutoff,
      minimalPeriod: minPeriod,
      maximalPeriod: maxPeriod,
    });

    this.rasterLength = rasterLength;
    this.rasterStep = maxRaster / rasterLength;
    this.maxRasterValue = maxRaster;
    this.minParameterValue = minParam;
    this.maxParameterValue = maxParam;
    this.parameterResolution = (rasterLength - 1) / (maxParam - minParam);
    this.raster = new Array<number>(rasterLength).fill(0);

    const cm = componentTripleMnemonic(bc, qc, tc);
    this.mnemonicValue = `cswing(${rasterLength}, ${maxRaster}, ${minParam}, ${maxParam}, ${hpCutoff}${cm})`;
    this.mnemonicSP = `cswing-sp(${hpCutoff}${cm})`;
    this.descriptionValue = 'Corona swing position ' + this.mnemonicValue;
    this.descriptionSP = 'Corona swing position scalar ' + this.mnemonicSP;
  }

  public isPrimed(): boolean { return this.c.isPrimed(); }

  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.CoronaSwingPosition,
      this.mnemonicValue,
      this.descriptionValue,
      [
        { mnemonic: this.mnemonicValue, description: this.descriptionValue },
        { mnemonic: this.mnemonicSP, description: this.descriptionSP },
      ],
    );
  }

  /** Feeds the next sample and returns the heatmap column plus the current SwingPosition. */
  public update(sample: number, time: Date): [Heatmap, number] {
    if (Number.isNaN(sample)) {
      return [
        Heatmap.newEmptyHeatmap(time, this.minParameterValue, this.maxParameterValue, this.parameterResolution),
        Number.NaN,
      ];
    }

    const primed = this.c.update(sample);

    if (!this.isStarted) {
      this.samplePrevious = sample;
      this.isStarted = true;
      return [
        Heatmap.newEmptyHeatmap(time, this.minParameterValue, this.maxParameterValue, this.parameterResolution),
        Number.NaN,
      ];
    }

    // Bandpass InPhase filter at dominant cycle median period.
    const omega = 2 * Math.PI / this.c.dominantCycleMedian;
    const beta2 = Math.cos(omega);
    const gamma2 = 1 / Math.cos(omega * 2 * BP_DELTA);
    const alpha2 = gamma2 - Math.sqrt(gamma2 * gamma2 - 1);
    const bandPass = 0.5 * (1 - alpha2) * (sample - this.samplePrevious2)
      + beta2 * (1 + alpha2) * this.bandPassPrevious
      - alpha2 * this.bandPassPrevious2;

    const quadrature2 = (bandPass - this.bandPassPrevious) / omega;

    this.bandPassPrevious2 = this.bandPassPrevious;
    this.bandPassPrevious = bandPass;
    this.samplePrevious2 = this.samplePrevious;
    this.samplePrevious = sample;

    // 60° lead.
    const lead60 = LEAD60_COEF_BP * this.bandPassPrevious2 + LEAD60_COEF_Q * quadrature2;

    let [lowest, highest] = appendRolling(this.leadList, MAX_LEAD_LIST_COUNT, lead60);

    let position = highest - lowest;
    if (position > 0) {
      position = (lead60 - lowest) / position;
    }

    [lowest, highest] = appendRolling(this.positionList, MAX_POSITION_LIST_COUNT, position);
    highest -= lowest;

    let width = 0.15 * highest;
    if (highest > WIDTH_HIGH_THRESHOLD) {
      width = WIDTH_NARROW;
    }

    this.swingPosition = (this.maxParameterValue - this.minParameterValue) * position + this.minParameterValue;

    const positionScaledToRasterLength = Math.round(position * this.rasterLength);
    const positionScaledToMaxRasterValue = position * this.maxRasterValue;

    for (let i = 0; i < this.rasterLength; i++) {
      let value = this.raster[i];

      if (i === positionScaledToRasterLength) {
        value *= RASTER_BLEND_HALF;
      } else {
        let argument = positionScaledToMaxRasterValue - this.rasterStep * i;
        if (i > positionScaledToRasterLength) {
          argument = -argument;
        }
        if (width > 0) {
          value = RASTER_BLEND_HALF * (Math.pow(argument / width, RASTER_BLEND_EXPONENT) + RASTER_BLEND_HALF * value);
        }
      }

      if (value < 0) {
        value = 0;
      } else if (value > this.maxRasterValue) {
        value = this.maxRasterValue;
      }

      if (highest > WIDTH_HIGH_SATURATE) {
        value = this.maxRasterValue;
      }

      if (Number.isNaN(value)) {
        value = 0;
      }

      this.raster[i] = value;
    }

    if (!primed) {
      return [
        Heatmap.newEmptyHeatmap(time, this.minParameterValue, this.maxParameterValue, this.parameterResolution),
        Number.NaN,
      ];
    }

    const values = new Array<number>(this.rasterLength);
    let valueMin = Number.POSITIVE_INFINITY;
    let valueMax = Number.NEGATIVE_INFINITY;
    for (let i = 0; i < this.rasterLength; i++) {
      const v = this.raster[i];
      values[i] = v;
      if (v < valueMin) valueMin = v;
      if (v > valueMax) valueMax = v;
    }

    const heatmap = Heatmap.newHeatmap(
      time, this.minParameterValue, this.maxParameterValue, this.parameterResolution,
      valueMin, valueMax, values,
    );

    return [heatmap, this.swingPosition];
  }

  public updateScalar(sample: Scalar): IndicatorOutput { return this.updateEntity(sample.time, sample.value); }
  public updateBar(sample: Bar): IndicatorOutput { return this.updateEntity(sample.time, this.barComponentFunc(sample)); }
  public updateQuote(sample: Quote): IndicatorOutput { return this.updateEntity(sample.time, this.quoteComponentFunc(sample)); }
  public updateTrade(sample: Trade): IndicatorOutput { return this.updateEntity(sample.time, this.tradeComponentFunc(sample)); }

  private updateEntity(time: Date, sample: number): IndicatorOutput {
    const [heatmap, sp] = this.update(sample, time);
    const s = new Scalar();
    s.time = time;
    s.value = sp;
    return [heatmap, s];
  }
}

/** Appends v to the list, drops the oldest once len reaches maxCount, returns [lowest, highest]. */
function appendRolling(list: number[], maxCount: number, v: number): [number, number] {
  if (list.length >= maxCount) {
    list.shift();
  }
  list.push(v);

  let lowest = v;
  let highest = v;
  for (let i = 0; i < list.length; i++) {
    const x = list[i];
    if (x < lowest) lowest = x;
    if (x > highest) highest = x;
  }
  return [lowest, highest];
}
